### PR TITLE
fix(core): fix phantom select query errors caused by concurrent partition squashing

### PR DESCRIPTION
**Pull request by ideoma**
Phantom query error reading var length column with the error message

```
dataLim < dataLo + size
```

caused by TableWriter squashing the partition opened for reading.

This PR introduces a strict test framework with the class `PartitionOverwriteControl` that tracks all open for read partitions and checks them every time a partition is about to be overwritten by TableWriter or O3PartitionPurgeJob. It is enabled in tests and disabled by default in the release configuration.

The bug that caused the squashing into the incorrect position was found and fixed.

This PR also changes `PARANOIA_FD_MODE` back to false after leaving it on in #5048

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 96 / 104 (92.31%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/PartitionOverwriteControl.java|58|66|87.88%|
|:large_blue_circle:|io/questdb/cairo/TableReader.java|12|12|100.00%|
|:large_blue_circle:|io/questdb/cairo/TxReader.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cairo/O3PartitionPurgeJob.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/cairo/pool/ReaderPool.java|7|7|100.00%|

File: core/src/main/java/io/questdb/PropertyKey.java [hunk 1]
```diff
@@ -473,6 +473,7 @@ public enum PropertyKey implements ConfigPropertyKey {
     CAIRO_REPEAT_MIGRATION_FROM_VERSION("cairo.repeat.migration.from.version"),
     CAIRO_O3_LAST_PARTITION_MAX_SPLITS("cairo.o3.last.partition.max.splits"),
     CAIRO_O3_PARTITION_SPLIT_MIN_SIZE("cairo.o3.partition.split.min.size"),
+    CAIRO_O3_PARTITION_OVERWRITE_CONTROL_ENABLE("cairo.o3.partition.overwrite.control.enable"),
```
Comment (puzpuzpuz) [inline core/src/main/java/io/questdb/PropertyKey.java:476]: nit: in all other boolean props we use `.enabled`, not `enable`.

File: core/src/main/java/io/questdb/cairo/TableWriter.java [hunk 2]
```diff
@@ -7619,120 +7620,134 @@ private void squashSplitPartitions(long timestampMin, long timestampMax, int max
     }
 
     private void squashSplitPartitions(final int partitionIndexLo, final int partitionIndexHi, final int optimalPartitionCount, boolean force) {
+        if (partitionIndexHi <= partitionIndexLo + Math.max(1, optimalPartitionCount)) {
+            // Noting to do
```
Comment (puzpuzpuz) [inline core/src/main/java/io/questdb/cairo/TableWriter.java:7624]: nit:
```suggestion
            // Nothing to do
```

File: core/src/test/java/io/questdb/test/ServerMainTest.java [hunk 3]
```diff
@@ -589,7 +589,8 @@ public void testShowParameters() throws Exception {
                                     "http.min.server.keep.alive\tQDB_HTTP_MIN_SERVER_KEEP_ALIVE\ttrue\tdefault\tfalse\tfalse\n" +
                                     "http.min.connection.string.pool.capacity\tQDB_HTTP_MIN_CONNECTION_STRING_POOL_CAPACITY\t2\tdefault\tfalse\tfalse\n" +
                                     "http.min.connection.pool.initial.capacity\tQDB_HTTP_MIN_CONNECTION_POOL_INITIAL_CAPACITY\t2\tdefault\tfalse\tfalse\n" +
-                                    "http.min.multipart.idle.spin.count\tQDB_HTTP_MIN_MULTIPART_IDLE_SPIN_COUNT\t0\tdefault\tfalse\tfalse\n"
+                                    "http.min.multipart.idle.spin.count\tQDB_HTTP_MIN_MULTIPART_IDLE_SPIN_COUNT\t0\tdefault\tfalse\tfalse\n" +
+                                    "airo.o3.partition.overwrite.control.enable\tQDB_CAIRO_O3_PARTITION_OVERWRITE_CONTROL_ENABLE\tfalse\tdefault\tfalse\tfalse\n"
```
Comment (puzpuzpuz) [inline core/src/test/java/io/questdb/test/ServerMainTest.java:593]:
```suggestion
                                    "cairo.o3.partition.overwrite.control.enable\tQDB_CAIRO_O3_PARTITION_OVERWRITE_CONTROL_ENABLE\tfalse\tdefault\tfalse\tfalse\n"
```

File: core/src/main/java/io/questdb/cairo/pool/ReaderPool.java [hunk 4]
```diff
@@ -66,26 +73,30 @@ public static class R extends TableReader implements PoolTenant<R> {
         private final ReaderListener readerListener;
         private Entry<R> entry;
         private AbstractMultiTenantPool<R> pool;
+        boolean active;
```
Comment (puzpuzpuz) [inline core/src/main/java/io/questdb/cairo/pool/ReaderPool.java:76]: This field seems to be unused. Do we need it?

File: core/src/main/java/io/questdb/cairo/PartitionOverwriteControl.java [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+import io.questdb.std.ConcurrentHashMap;
+import io.questdb.std.LongList;
+import io.questdb.std.ObjList;
+
+// When enabled, this class tracks TableReader usage of partitions.
+// The TableWriter and O3PartitionPurgeJob check if the particular partition version is in use to prevent partition overwrite bugs.
+// This is supposed to be disabled in production runs and only enabled in tests or on special debugging occasions via configuration.
+public class PartitionOverwriteControl {
+    private static final Log LOG = LogFactory.getLog(PartitionOverwriteControl.class);
+    boolean enabled;
```
Comment (puzpuzpuz) [inline core/src/main/java/io/questdb/cairo/PartitionOverwriteControl.java:38]: nit: the fields and the `ReaderPartitionUsage` class can be private.

**Comment by ideoma** [comment]
> @ideoma could you point me at the exact place in `TableWriter#squashSplitPartitions()` where the bugfix landed?

The bug was that after finding correct `targetPartitionIndex` the squashing ignored it and always chose the target partition to be `partitionIndexLo`

File: core/src/main/java/io/questdb/cairo/pool/ReaderPool.java (hunk 4 above)
Comment (ideoma) [reply core/src/main/java/io/questdb/cairo/pool/ReaderPool.java]: removed

File: core/src/main/java/io/questdb/PropertyKey.java (hunk 1 above)
Comment (ideoma) [reply core/src/main/java/io/questdb/PropertyKey.java]: changed to `enabled`

File: core/src/main/java/io/questdb/cairo/TableWriter.java (hunk 2 above)
Comment (ideoma) [reply core/src/main/java/io/questdb/cairo/TableWriter.java]: changed as suggested

**Review by puzpuzpuz — APPROVED** [review]
Nice idea with the `PartitionOverwriteControl` object!

**Review by puzpuzpuz — APPROVED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 185 / 196 (94.39%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/PartitionOverwriteControl.java|57|65|87.69%|
|:large_blue_circle:|io/questdb/cairo/CairoConfigurationWrapper.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|80|82|97.56%|
|:large_blue_circle:|io/questdb/cairo/TableReader.java|12|12|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/TxReader.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/DefaultCairoConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/O3PartitionPurgeJob.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Files.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/pool/ReaderPool.java|5|5|100.00%|

**Pull request merged**